From 0a9f9a7410681e55362f8311537ebc7be9ad0fbe Mon Sep 17 00:00:00 2001
From: Sascha Hauer <s.hauer@pengutronix.de>
Date: Fri, 5 Mar 2021 09:27:22 +0100
Subject: [PATCH] crypto: digest: use crypto_memneq()

When verifying a digest it is important not to leak timing information
through memcmp(). Use crypto_memneq() instead.

Signed-off-by: Sascha Hauer <s.hauer@pengutronix.de>
---
 crypto/digest.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/crypto/digest.c b/crypto/digest.c
index d23245e15f..621d384168 100644
--- a/crypto/digest.c
+++ b/crypto/digest.c
@@ -22,6 +22,7 @@
 #include <errno.h>
 #include <module.h>
 #include <linux/err.h>
+#include <crypto.h>
 #include <crypto/internal.h>
 
 static LIST_HEAD(digests);
@@ -47,8 +48,10 @@ int digest_generic_verify(struct digest *d, const unsigned char *md)
 	if (ret)
 		goto end;
 
-	ret = memcmp(md, tmp, len);
-	ret = ret ? -EINVAL : 0;
+	if (crypto_memneq(md, tmp, len))
+		ret = -EINVAL;
+	else
+		ret = 0;
 end:
 	free(tmp);
 	return ret;
